nebula_ros/src/velodyne/decoder_wrapper.cpp

```diff
@@ -0,0 +1,248 @@
+  const std::vector<rclcpp::Parameter> & p)
+{
+  using rcl_interfaces::msg::SetParametersResult;
+
+  std::string calibration_path = "";
+
+  bool got_any = get_param(p, "calibration_file", calibration_path);
+  if (!got_any) {
+    return rcl_interfaces::build<SetParametersResult>().successful(true).reason("");
+  }
+
+  if (!std::filesystem::exists(calibration_path)) {
+    auto result = SetParametersResult();
+    result.successful = false;
+    result.reason =
+      "The given calibration path does not exist, ignoring: '" + calibration_path + "'";
+    return result;
+  }
+
+  auto get_calibration_result = GetCalibrationData(calibration_path);
+  if (!get_calibration_result.has_value()) {
+    auto result = SetParametersResult();
+    result.successful = false;
+    result.reason =
+      (std::stringstream() << "Could not change calibration file to '" << calibration_path
+                           << "': " << get_calibration_result.error())
+        .str();
+    return result;
+  }
+
+  OnCalibrationChange(get_calibration_result.value());
+  RCLCPP_INFO_STREAM(
+    logger_, "Changed calibration to '" << calibration_cfg_ptr_->calibration_file << "'");
+  return rcl_interfaces::build<SetParametersResult>().successful(true).reason("");
+}
+
+VelodyneDecoderWrapper::get_calibration_result_t VelodyneDecoderWrapper::GetCalibrationData(
+  const std::string & calibration_file_path)
+{
+  auto calib = std::make_shared<drivers::VelodyneCalibrationConfiguration>();
+
+  // If downloaded data did not exist or could not be loaded, fall back to a generic file.
+  // If that file does not exist either, return an error code
+  if (!std::filesystem::exists(calibration_file_path)) {
+    RCLCPP_ERROR(logger_, "No calibration data found.");
+    return nebula::Status(Status::INVALID_CALIBRATION_FILE);
+  }
+
+  // Try to load the existing fallback calibration file. Return an error if this fails
+  auto status = calib->LoadFromFile(calibration_file_path);
+  if (status != Status::OK) {
+    RCLCPP_ERROR_STREAM(
+      logger_, "Could not load calibration file at '" << calibration_file_path << "'");
+    return status;
+  }
+
+  // Return the fallback calibration file
+  calib->calibration_file = calibration_file_path;
+  return calib;
+}
+
+void VelodyneDecoderWrapper::ProcessCloudPacket(
+  std::unique_ptr<nebula_msgs::msg::NebulaPacket> packet_msg)
+{
+  // Accumulate packets for recording only if someone is subscribed to the topic (for performance)
+  if (
+    hw_interface_ && (packets_pub_->get_subscription_count() > 0 ||
+                      packets_pub_->get_intra_process_subscription_count() > 0)) {
+    if (current_scan_msg_->packets.size() == 0) {
+      current_scan_msg_->header.stamp = packet_msg->stamp;
+    }
+
+    velodyne_msgs::msg::VelodynePacket velodyne_packet_msg{};
+    velodyne_packet_msg.stamp = packet_msg->stamp;
+    std::copy(packet_msg->data.begin(), packet_msg->data.end(), velodyne_packet_msg.data.begin());
```

Comment:
Please correct me if I am wrong. Is the copy necessary?

Reply:
Sadly VelodynePacket uses `std::array` while NebulaPacket uses `std::vector`. If both were `vector`, we could indeed swap them instead but as it is, this copy is necessary.

Reply:
Thanks!